packetbeat/protos/dns/dns.go

```diff
@@ -383,240 +392,229 @@ func (dns *Dns) expireTransaction(t *DnsTransaction) {
 	mapStrArray := make([]common.MapStr, len(records))
-	for i, r := range records {
+	for i, rr := range records {
+		rrHeader := rr.Header()
+		rrType := rrHeader.Rrtype
+
 		mapStr := common.MapStr{
-			"name":  nameToString(r.Name),
-			"type":  dnsTypeToString(r.Type),
-			"class": dnsClassToString(r.Class),
-			"ttl":   r.TTL,
+			"name":  rrHeader.Name,
+			"type":  dnsTypeToString(rrType),
+			"class": dnsClassToString(rrHeader.Class),
+			"ttl":   strconv.FormatInt(int64(rrHeader.Ttl), 10),
 		}
 		mapStrArray[i] = mapStr
 
-		switch r.Type {
+		switch x := rr.(type) {
 		default:
-			// We don't have special handling for this type so use the same
-			// encoding used for names to output the raw data for this type.
-			mapStr["data"] = nameToString(r.Data)
-		case layers.DNSTypeA, layers.DNSTypeAAAA:
-			mapStr["data"] = r.IP.String()
-		case layers.DNSTypeSOA:
-			mapStr["rname"] = nameToString(r.SOA.RName)
-			mapStr["serial"] = r.SOA.Serial
-			mapStr["refresh"] = r.SOA.Refresh
-			mapStr["retry"] = r.SOA.Retry
-			mapStr["expire"] = r.SOA.Expire
-			mapStr["minimum"] = r.SOA.Minimum
-			mapStr["data"] = nameToString(r.SOA.MName)
-		case layers.DNSTypeMX:
-			mapStr["preference"] = r.MX.Preference
-			mapStr["data"] = nameToString(r.MX.Name)
-		case layers.DNSTypeSRV:
-			mapStr["priority"] = r.SRV.Priority
-			mapStr["weight"] = r.SRV.Weight
-			mapStr["port"] = r.SRV.Port
-			mapStr["data"] = nameToString(r.SRV.Name)
-		case layers.DNSTypeCNAME:
-			mapStr["data"] = nameToString(r.CNAME)
-		case layers.DNSTypePTR:
-			mapStr["data"] = nameToString(r.PTR)
-		case layers.DNSTypeNS:
-			mapStr["data"] = nameToString(r.NS)
+			// We don't have special handling for this type
+			logp.Debug("dns", "Unsupported RR type %s", dnsTypeToString(rrType))
+		case *mkdns.A:
+			mapStr["data"] = x.A.String()
+		case *mkdns.AAAA:
+			mapStr["data"] = x.AAAA.String()
+		case *mkdns.CNAME:
+			mapStr["data"] = x.Target
+		case *mkdns.MX:
+			mapStr["preference"] = x.Preference
+			mapStr["data"] = x.Mx
+		case *mkdns.NS:
+			mapStr["data"] = x.Ns
+		case *mkdns.PTR:
+			mapStr["data"] = x.Ptr
+		case *mkdns.RFC3597:
+			// Miekg/dns lib doesn't handle this type
+			logp.Debug("dns", "Unknown RR type %s", dnsTypeToString(rrType))
+			mapStr["data"] = x.Rdata
+		case *mkdns.SOA:
+			mapStr["rname"] = x.Mbox
+			mapStr["serial"] = x.Serial
+			mapStr["refresh"] = x.Refresh
+			mapStr["retry"] = x.Retry
+			mapStr["expire"] = x.Expire
+			mapStr["minttl"] = x.Minttl
+			mapStr["data"] = x.Ns
+		case *mkdns.SRV:
+			mapStr["priority"] = x.Priority
+			mapStr["weight"] = x.Weight
+			mapStr["port"] = x.Port
+			mapStr["data"] = x.Target
+		case *mkdns.TXT:
+			mapStr["data"] = strings.Join(x.Txt, " ")
 		}
 	}
 
 	return mapStrArray
 }
 
-// dnsQuestionToString converts a DNSQuestion to a string.
-func dnsQuestionToString(q layers.DNSQuestion) string {
-	name := nameToString(q.Name)
-	if len(name) == 0 {
-		name = "Root"
```

Comment:
If the domain name == 0, then it is set to "." [See func UnpackDomainName](https://github.com/miekg/dns/blob/master/msg.go#L397-L399)